lang/funcs/core/coresys/cpucount_fact.go

```diff
@@ -0,0 +1,210 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package coresys
+
+import (
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/purpleidea/mgmt/lang/funcs/facts"
+	"github.com/purpleidea/mgmt/lang/types"
+	"github.com/purpleidea/mgmt/util/socketset"
+
+	errwrap "github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	rtmGrps         = 0x1 // make me a multicast reciever
+	socketFile      = "pipe.sock"
+	cpuDevpathRegex = "/devices/system/cpu/cpu[0-9]"
+)
+
+func init() {
+	facts.ModuleRegister(moduleName, "cpu_count", func() facts.Fact { return &CPUCountFact{} }) // must register the fact and name
+}
+
+// CPUCountFact is a fact that returns the current CPU count.
+type CPUCountFact struct {
+	init      *facts.Init
+	closeChan chan struct{}
+}
+
+// Info returns static typing info about what the fact returns.
+func (obj *CPUCountFact) Info() *facts.Info {
+	return &facts.Info{
+		Output: types.NewType("int"),
+	}
+}
+
+// Init run startup code for this fact. Initializes the closeChan
+// and sets the facts.Init variable.
+func (obj *CPUCountFact) Init(init *facts.Init) error {
+	obj.init = init
+	obj.closeChan = make(chan struct{})
+	return nil
+}
+
+// Stream returns the changing values that this fact has over time.
+// It will first poll sysfs to get the initial cpu count, and then
+// receieves UEvents from the kernel as CPUs are added/removed.
+func (obj CPUCountFact) Stream() error {
+	// waitgroup for netlink recieve goroutine
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+
+	defer close(obj.init.Output) // signal when we're done
```

Comment:
This should happen last. IOW should be at the top of this function. We only want to tell the func engine that we're done, when we're done (after the waitgroup Wait returns)